src/coreclr/utilcode/executableallocator.cpp

```diff
@@ -276,28 +326,53 @@ void ExecutableAllocator::RemoveCachedMapping()
         g_fatalErrorHandler(COR_E_EXECUTIONENGINE, W("Releasing the RW mapping failed"));
     }
 
-    m_cachedMapping = NULL;
+    m_cachedMapping[index - 1] = NULL;
 #endif // ENABLE_CACHED_MAPPINGS
 }
 
+#ifdef ENABLE_CACHED_MAPPINGS
+size_t ExecutableAllocator::FindOverlappingCachedMapping(BlockRX* pBlock)
+{
+    for (size_t index = 0; index < EXECUTABLE_ALLOCATOR_CACHE_SIZE; index++)
+    {
+        BlockRW* cachedMapping = m_cachedMapping[index];
+        if (cachedMapping != NULL)
+        {
+            // In case the cached mapping maps the region being released, it needs to be removed
+            if ((pBlock->baseRX <= cachedMapping->baseRX) && (cachedMapping->baseRX < ((BYTE*)pBlock->baseRX + pBlock->size)))
+            {
+                return index + 1;
+            }
+        }
+    }
+    return 0;
+}
+#endif
+
 void ExecutableAllocator::UpdateCachedMapping(BlockRW* pBlock)
 {
     LIMITED_METHOD_CONTRACT;
 #ifdef ENABLE_CACHED_MAPPINGS
-    if (m_cachedMapping != pBlock)
+    for (size_t index = 0; index < EXECUTABLE_ALLOCATOR_CACHE_SIZE; index++)
     {
-        if (m_cachedMapping != NULL)
+        if (pBlock == m_cachedMapping[index])
         {
-            RemoveCachedMapping();
+            // Move the found mapping to the front - note the overlapping memory, use memmove.
+            memmove(&m_cachedMapping[1], &m_cachedMapping[0], sizeof(m_cachedMapping[0]) * index);
+            m_cachedMapping[0] = pBlock;
+            return;
```

Comment:
It is missing adding the refcount when you return here.

Reply:
It doesn't need a ref count in this situation. In this situation it is finding the value in the cache, and like the old code doesn't need to adjust any ref counts.

Reply:
I am sorry, you are right, I've misread my original code.